003 커뮤니티

shCHO9801 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 클라이밍 관련 기능 외에 커뮤니티(게시글, 미디어) 기능 미구현
- 게시글 관련 엔티티와 Repository 미구현으로, 게시글에 여러 첨부 미디어(사진/동영상)를 처리할 수 없는 상태
- 게시글 CRUD API 미구현 (생성, 수정, 조회, 삭제 기능 모두 누락)
- 게시글 조회 시 정렬 및 페이지네이션 기능 미지원
- 게시글 관련 데이터(첨부 미디어 URL)가 게시글 엔티티에 직접 저장되어 데이터 구조가 난잡함
- 관련 기능에 대한 유닛테스트 및 통합 테스트 미작성

**TO-BE**
- 게시글(Post) 및 미디어(PostMedia) 엔티티 구현
  - 게시글(Post) 엔티티는 제목, 내용, 작성자(User), 클라이밍장(ClimbingGym) 정보, 생성/수정 일시 등을 포함
  - 미디어(PostMedia) 엔티티는 미디어 URL, 미디어 타입(MediaType: IMAGE, VIDEO)을 가지며, 게시글과 ManyToOne 관계를 가짐
- Repository 구현
  - PostRepository와 PostMediaRepository를 구현하여 각각 게시글과 첨부 미디어의 CRUD를 지원
- 서비스 레이어 구현
  - PostService를 구현하여 게시글 생성 시, 로그인한 사용자의 아이디를 기준으로 게시글과 함께 첨부 미디어(PostMedia)를 별도의 테이블에 저장하도록 구성
  - 게시글 조회 시에는 페이지네이션 및 최신 순(예: createdAt 내림차순) 기능을 적용하여 반환
  - 게시글 수정 및 삭제 시에는 작성자와 요청자 일치 여부를 검사하여 권한 없는 수정/삭제를 방지
  - AWS S3 연동은 추후 계획이며, 현재는 미디어 URL을 전달받아 저장하는 형태로 구현
- 컨트롤러 구현
  - PostController에서는 JWT 토큰을 이용하여 사용자 인증 후 게시글 CRUD API를 제공
  - JWT 토큰은 "Bearer {token}" 형식으로 전달되며, 컨트롤러 내에서 추출 및 사용자 ID 확인 후 서비스 호출
  - 생성, 조회(페이지네이션 포함), 수정, 삭제 API 구현
- DTO 구성
  - CreatePostRequest: 제목, 내용, 클라이밍장 ID, 최대 10개까지의 미디어(PostMediaRequest) 목록을 입력받도록 정의 (미디어 갯수는@Size(max=10) 등으로 검증)
  - CreatePostResponse: 게시글 생성 결과 (게시글 ID 및 성공 메시지 등)
  - UpdatePostRequest: 제목과 내용 수정에 필요한 데이터

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- 유닛 테스트
  - Service 단에서 PostService의 createPost, getPost, updatePost, deletePost 등의 메서드에 대해
정상/비정상 시나리오(예: 사용자 미존재, 클라이밍장 미존재, 권한 없는 수정/삭제)를 검증하는 테스트 코드 작성
  - Repository와의 상호작용(예: 첨부 미디어 저장)이 올바르게 호출되는지 검증
- 통합 테스트
  - PostController를 대상으로 실제 API 호출을 통해 게시글 생성 및 조회, 수정, 삭제가 올바르게 동작하는지 확인
  - JWT 토큰을 통한 인증(Authorization 헤더에서 Bearer 토큰 처리) 및 페이지네이션 적용된 게시글 조회 기능 테스트
- [ ] API 테스트

shCHO9801 committed:
feat : Post, PostMedia 엔티티 및 레포지토리 구현

shCHO9801 committed:
feat : Post, PostMedia DTO 추가 및 로직 추가

shCHO9801 committed:
test : post 인테그레이션, 유닛 테스트 완료

shCHO9801 committed:
feat : Comment 설계 및 구현
- Comment Entity, Repository 구현
- 댓글관련 생성, 수정, 삭제 기능 추가

shCHO9801 committed:
test : comment 유닛테스트 작성
- 댓글 생성
  - 성공
  - 실패
    - 유저 미존재
    - 게시글 미존재
- 게시글 수정
  - 성공
  - 실패 - 권한 없음
- 게시글 삭제
  - 성공
  - 실패 - 권한 없음
- 게시글 목록 조회

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/comment/controller/CommentController.java:
```diff
@@ -0,0 +1,122 @@
+  private final JwtTokenProvider provider;
+  private final CommentService commentService;
+
+  @PostMapping("/{postId}")
```
api path는 /api/posts/{postId}/comments 가 좀더 적합할 것 같아요.

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/comment/dto/UpdateCommentRequest.java:
```diff
@@ -0,0 +1,14 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class UpdateCommentRequest {
+  private String comment;
```
`CreateCommentRequest `에서는 content라는 표현을 사용한 것 같은데 여기도 동일하게 content로 쓰면 좋을 것 같아요

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/comment/service/CommentService.java:
```diff
@@ -0,0 +1,93 @@
+        .orElseThrow(() -> new CustomException(USER_NOT_FOUND));
+  }
+
+  private Comment buildComment(User user, Post post, CreateCommentRequest request) {
+    return Comment.builder()
+        .content(request.getContent())
+        .post(post)
+        .user(user)
+        .build();
+  }
```
요 코드를 여기에 두셔도 되고 Comment 하위에 두는 경우도 많습니다. 
Comment 하위에 두실 경우 api controller에서 commentService layer로 내려올때 CreateCommentRequest를 그대로 넘기지 마시고 CreateCommentRequest안의 필요 필드만 별도로 넘기시면 좋을 것 같아요. 

~~~java
private Comment buildComment(User user, Post post, String content) {
    return Comment.builder()
        .content(content)
        .post(post)
        .user(user)
        .build();
  }
~~~

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/comment/service/CommentService.java:
```diff
@@ -0,0 +1,93 @@
+    Comment comment = commentRepository.findById(commentId)
+        .orElseThrow(() -> new CustomException(ErrorCode.COMMENT_NOT_FOUND));
+
+    if(!postRepository.existsById(comment.getPost().getId())) {
+      throw new CustomException(ErrorCode.UNAUTHORIZED_ACTION);
+    }
```
post id가 없는 경우는 post가 먼저 삭제된 경우일까요? 이 경우 `UNAUTHORIZED_ACTION` 요 exception을 던지시는 이유가 궁금합니다.

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/posting/service/PostService.java:
```diff
@@ -0,0 +1,130 @@
+      throw new CustomException(UNAUTHORIZED_ACTION);
+    }
+
+    postRepository.delete(post);
```
post 삭제시에 하위 comment는 따로 삭제를 안하나요?
comment와 post 사이 fk가 걸려있을것 같은데 post만 삭제되면 연결된 post가 없는 comment들이 생길것 같아서요

mingming-mentor commented on src/main/java/com/github/shCHO9801/climbing_record_app/community/posting/service/PostService.java:
```diff
@@ -0,0 +1,130 @@
+  private Post setUpdatePost(UpdatePostRequest request, Post post) {
+    post.setTitle(request.getTitle());
+    post.setContent(request.getContent());
+    post.setUpdatedAt(LocalDateTime.now());
```
Post entity에 @PreUpdate 가 있어서 요 부분은 없어도 되지 않을까 싶어요

mingming-mentor reviewed (approved):
수고 많으셨습니다

shCHO9801 committed:
fix : 코드 내용 피드백 반영
- CommentController path 변경
- Comment dto 내용 일관되게 수정 및 build 메서드 추가
- CommentService : 유저 인증 부분 수정 및 에러 코드 변경

shCHO9801 committed:
test : Comment 인테그레이션 테스트

shCHO9801 committed:
fix : Comment 인테그레이션 테스트 수정
- JSON 응답을 문자열 조작이 아닌 JsonNode를 활용하여 추출
